### ISSUE TITLE
Interfaces for laser-plasma codes

### ISSUE THREAD

**Issue by valenpe7** [valenpe7/laser_bcs#4]
The laser_bcs library is supposed to be integrated with electromagnetic codes. To simplify the communication between the library functions with other codes, it is necessary to write some interfaces. Up to now, there is only one code supported (Epoch: https://cfsa-pmw.warwick.ac.uk/). Add more interfaces (e. g. Smilei, Warp, ...)

**Comment by ax3l** [comment]
Do you want tho add bindings to [PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/)? ;-)

**Comment by valenpe7** [comment]
@ax3l Yes, I have it in my mind as well! :-) But first, I'd like to rewrite output routines (issue [#13](https://github.com/valenpe7/laser_bcs/issues/13)). Currently, the code uses MPI binary file, so one needs additional methods to retrieve the data. I haven't tried PIConGPU yet.. is there any easy way how to initialize a custom laser from file?

**Comment by ax3l** [comment]
File dumping (yay, you want to use openPMD as well) is a good idea for development. Have you seen [openPMD-api](https://github.com/openPMD/openPMD-api) yet? For code coupling, dumping the *whole* field is ok-ish but can be very storage and I/O intensive at scale. Just keep this in mind, parallel HDF5 does not perform too well [at scale](https://arxiv.org/abs/1706.00522) and you might want to consider in-transit ("staged") methods later on (also very much possible with openPMD).

In PIConGPU we currently initialize laser profiles two-and-a-half ways.
The default one "laser.param" initializes the em-field from a border or plane directly. An other way is to use the "fieldBackground.param" to define functions `E(x,y,z,t)` and `B(x,y,z,t)`. In the same file one could also define a `J(x,y,z,t)` for a custom antenna.

If you want to hack on the boundary conditions there are two ways, the quick and the fast one. The quick way which would allow you to use your (host-side) C++ routines is just to synchronize the fields to  the host ([`FieldE.synchronize()`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/cc80e36b1fc2b02e8ce7a5a21450c8fc0beb0217/include/picongpu/fields/FieldE.hpp#L93)), manipulate the `getHostDataBox()` like any other C++ array and push it back ([`FieldE.syncToDevice()`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/cc80e36b1fc2b02e8ce7a5a21450c8fc0beb0217/include/picongpu/fields/FieldE.hpp#L95)). This gives you quick results but is limited by the latency and bandwidth of PCI-E / nvlink.

The fast way: we [write a kernel](https://github.com/ComputationalRadiationPhysics/picongpu/blob/cc80e36b1fc2b02e8ce7a5a21450c8fc0beb0217/include/picongpu/fields/FieldManipulator.kernel#L39-L149) that updates the field directly on the device - probably also not too to change in your code as well :)

If you want to come around and visit our lab to work with us for a week or two just drop us a line, Prague and Dresden are really close :-)
